netbench-orchestrator/src/russula_cli.rs

```diff
@@ -1,7 +1,177 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-#[allow(dead_code)]
+use crate::russula::netbench;
+use crate::russula::WorkflowState;
+use core::time::Duration;
+use russula::{
+    netbench::{client, server},
+    WorkflowBuilder,
+};
+use std::{collections::BTreeSet, net::SocketAddr};
+use structopt::StructOpt;
+use tracing::debug;
+use tracing_subscriber::EnvFilter;
+
 mod russula;
 
-fn main() {}
+/// This utility is a convenient CLI wrapper around Russula and can be used to launch
+/// different workflow.
+#[derive(StructOpt, Debug)]
+struct Opt {
+    /// Russula workers and coordinators must be polled to make progress
+    #[structopt(long, parse(try_from_str=parse_duration), default_value = "5s")]
+    poll_delay: Duration,
+
+    /// Select which Russula workflow to start
+    #[structopt(subcommand)]
+    workflow: RussulaWorkflow,
+}
+
+/// A list of different Russula workflow
+#[allow(clippy::enum_variant_names)]
+#[derive(StructOpt, Debug)]
+enum RussulaWorkflow {
+    NetbenchServerWorker {
+        /// The port on which the Worker should 'listen' on.
+        #[structopt(long)]
+        russula_port: u16,
+
+        #[structopt(flatten)]
+        ctx: netbench::ServerContext,
+    },
+    NetbenchClientWorker {
+        /// The port on which the Worker should 'listen' on.
+        #[structopt(long)]
+        russula_port: u16,
+
+        #[structopt(flatten)]
+        ctx: netbench::ClientContext,
+    },
+    NetbenchServerCoordinator {
+        /// The list of worker addresses which the Coordinator should
+        /// attempt to connect
+        #[structopt(long, required = true)]
+        russula_worker_addrs: Vec<SocketAddr>,
+    },
+    NetbenchClientCoordinator {
+        /// The list of worker addresses which the Coordinator should
+        /// attempt to connect
+        #[structopt(long)]
+        russula_worker_addrs: Vec<SocketAddr>,
+    },
+}
+
+#[tokio::main(flavor = "current_thread")]
```

Comment:
If there is a reason that we want this to be run on the current thread, could we document that? And if not, maybe just leave it up to the default?

Reply:
This simplifies the testing. The single thread wont be the bottleneck for the orchestrator and I dont want to re-test things with a multi-threaded runtime.

Reply:
Shouldn't the flavor of the runtime be largely opaque to the application? 

> and I dont want to re-test things with a multi-threaded runtime

Are there known issues with using the multi-threaded runtime? My understanding is that tokio enforces the same `Send + Sync` constraints independent of whether the runtime is actually using multiple threads or not, so we should be able to relax to the default without any worries about correctness.

Reply:
I don't expect there to be issues, but its also not worth re-testing with multi-thread.

Reply:
Can't we just use the unit tests/etc? Or is there some other large surface area to worry about?

Reply:
The goal of the unit tests is to test the workflow state machine rather than the networking.

As I said, I expect it would work but there is no need to introduce the complexity of multi-threaded since it's not going to be the bottleneck, especially since the IO is non-blocking. Multi-threading might actually reduce performance due to scheduling overhead.